### PR TITLE
Make catalog tables interactive

### PR DESCRIPTION
**Pull request by kcarver1**
### Description

This pull request is to address making the table interactive, mainly adding a zoom feature and additional markers for selected points. 

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 1]
```diff
@@ -160,13 +163,21 @@ def search(self, error_on_fail=False):
...
                                       unit='deg')
 
+            # show_rowselect = Bool(False).tag(sync=True)  # Flag to enable row selection boxes
+            # item_key = Unicode().tag(sync=True)  # Unique field to identify row for selection
+            # selected_rows = List().tag(sync=True)  # List of selected rows
             # adding in coords + Id's into table
+            self.table.item_key = 'id'
+            self.table.show_rowselect = True 
+            i = 0 
             for row in self.app._catalog_source_table:
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:173]: for loops can handle the counter for you (and can then get rid of the `i = 0` and ` i+= 1` lines

```suggestion
            for i, row in enumerate(self.app._catalog_source_table):
```

Note though that both of these approaches assumes the table is cleared between successive searches - is that the case?  If not, `id` could instead be something like `len(self.table._qtable)+1`.

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 2]
```diff
@@ -218,9 +229,23 @@ def search(self, error_on_fail=False):
         # markers are added to the viewer based on the table
         viewer.marker = {'color': 'red', 'alpha': 0.8, 'markersize': 5, 'fill': False}
         viewer.add_markers(table=catalog_results, use_skycoord=True, marker_name=self._marker_name)
-
+            
         return skycoord_table
 
+    def plot_selected_points(self,viewer):
```
Comment (pllim) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:235]: `viewer.add_markers` is expensive because you create a new table entry in data collection and all the relinking happens. If you want to interactively inspect stuff in existing table, maybe we can use the stuff from Markers plugin instead (not to be confused with Astrowidgets markers). @kecnry , since you wrote the Markers plugin, any ideas?

Comment (pllim) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:235]: I already showed it is possible to circle something, so maybe we can add additional circle overlay to selected data?

https://github.com/spacetelescope/jdaviz/pull/3092

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:235]: If the intention is to "highlight" the currently selected rows when the plugin is opened, then `PluginMark` would be appropriate (that's what is used for all plugin "live-previews").

**Comment by kcarver1** [comment]
Right now, I have the x and y pixel coordinates to be used to create markers. Additionally, I have created a class for a catalog specific marker. The markers are still not being added as expected however, so I pushed what I had to get some fresh eyes on the new code.

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 3]
```diff
@@ -233,7 +224,7 @@ def search(self, error_on_fail=False):
 
         self.table.item_key = 'id'
         self.table.show_rowselect = True
-        i = 0
+        i = len(skycoord_table) + 1
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:227]:
```suggestion
        i = len(skycoord_table)
```

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 4]
```diff
@@ -303,16 +292,19 @@ def _on_is_active_changed(self, *args):
 
     def plot_selected_points(self):
         selected_rows = self.table.selected_rows
+        if len(selected_rows) == 0 or 'x_coord' not in selected_rows[0]:
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:295]: when would x_coord not be populated?  Could it be in `selected_rows[0]` but not in another row?

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.vue [hunk 5]
```diff
@@ -67,6 +67,15 @@
             Search
           </plugin-action-button>
        </v-col>
+       <v-col>
+         <plugin-action-button
+            :results_isolated_to_plugin="true"
+            @click="zoom_in"
+            :spinner="spinner"
+          >
+            Zoom in
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.vue:76]:
```suggestion
            Zoom to Selected
```

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py (hunk 4 above)
Comment (javerbukh) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:295]: That was part of the bug I was seeing earlier. Rows were being added in the first `if catalog == 'SDSS'` without `x_coord` and then again later in the file with `x_coord`. These lines can safely be removed.

**Comment by gibsongreen** [comment]
I wasn't involved in the review process much so I have a few questions about the expected functionality:
1. From the UI, is it expected that when a selection of a particular source is made, that all are selected?
2. There is the `Max sources` input parameter, I would imagine the main use case being to prevent < 1000 sources. I was expecting inputing 1 to and expected the table to show only 1 result, using the Gaia catalog and the Example notebook data, I get 17 sources irregardless of changing this value (would we also want to expose this?)
3. Is it in scope for this ticket/do we want API expose functionality for `Zoom to Selected` & selecting a particular source from the table?
4. `get_catalog_source_results()` is accessible from the config, but I was expecting to see it in the directory for the plugin, is it desired/in scope to move this to the plugin instead?

**Comment by javerbukh** [comment]
2. I see what you mean but it looks like it's an issue with line https://github.com/spacetelescope/jdaviz/blob/a754a27a3bff4a282a493f19906d6f11b575bebd/jdaviz/configs/imviz/plugins/catalogs/catalogs.py#L179 so I think that would be outside the scope of this ticket and may need a follow-up.
3. Same as above, I hesitate for this ticket to become a catch all for Catalog's bug fixes + improvements when it was just meant to fix the functionality originally added by our intern.
4. Same as above.
5. I'm not sure why the table with `sky_centroid` would not load into the viewer, so another follow-up may be necessary.

To summarize, I think we need follow-up tickets for the following:
1. Bug: Gaia max sources are not being limited by the traitlet.
2. Expose API functionality for the Catalog Plugin and move `get_catalog_source_results()` to an appropriate location.
3. Enable `Zoom` and row select functionality for custom files.

**Comment by javerbukh** [comment]
Follow up tickets have been created: 

1. https://jira.stsci.edu/browse/JDAT-4770 
2. https://jira.stsci.edu/browse/JDAT-4771 
3. https://jira.stsci.edu/browse/JDAT-4772

**Comment by javerbukh** [comment]
Follow-up tickets have been created.

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 6]
```diff
@@ -228,6 +215,44 @@ def search(self, error_on_fail=False):
         filtered_pair_pixel_table = np.array(np.hsplit(filtered_table, 2))
         x_coordinates = np.squeeze(filtered_pair_pixel_table[0])
         y_coordinates = np.squeeze(filtered_pair_pixel_table[1])
+
+        self.table.item_key = 'id'
+        self.table.show_rowselect = True
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:220]: does this ever change (can it be moved to the init)?

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 7]
```diff
@@ -228,6 +215,44 @@ def search(self, error_on_fail=False):
...
+        i = len(skycoord_table)
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:221]: should this be the current length of the full table (not the current new results)?  Maybe instead of an iterator, the `id` assigned in `row_info` can just check the current length of `self.table`?

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 8]
```diff
@@ -236,11 +261,76 @@ def search(self, error_on_fail=False):
 
         self.number_of_results = len(catalog_results)
         # markers are added to the viewer based on the table
-        viewer.marker = {'color': 'red', 'alpha': 0.8, 'markersize': 5, 'fill': False}
+        viewer.marker = {'color': 'blue', 'alpha': 0.8, 'markersize': 30, 'fill': False}
         viewer.add_markers(table=catalog_results, use_skycoord=True, marker_name=self._marker_name)
-
         return skycoord_table
 
+    # unsure of whether or not I need to adjust this even more
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:268]: any follow-up on this?  If not, maybe we can just delete the comment?

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 9]
```diff
@@ -236,11 +261,76 @@ def search(self, error_on_fail=False):
...
+    def plot_selected_points(self):
+        selected_rows = self.table.selected_rows
+
+        x = [float(coord['x_coord']) for coord in selected_rows]
+        y = [float(coord['y_coord']) for coord in selected_rows]
+        self._get_mark(self.viewer.selected_obj).update_xy(getattr(x, 'value', x),
+                                                           getattr(y, 'value', y))
+
+    # this function will zoom into the image based on the selected points
+    def vue_zoom_in(self, *args, **kwargs):
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:300]:
```suggestion
    def vue_zoom_in(self, *args, **kwargs):
         """this function will zoom into the image based on the selected points"""
```

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 10]
```diff
@@ -236,11 +261,76 @@ def search(self, error_on_fail=False):
...
+        # extract max and min x and y coords; then add a bit of space around those as well
+        # zoom region will be rectangular
+        # Adding some padding around the zoom area
+        # padding_factor = 0.1  # 10% padding
+        # x_range = x_max - x_min + 100
+        # y_range = y_max - y_min + 100
+        # x_min -= x_range * padding_factor
+        # x_max += x_range * padding_factor
+        # y_min -= y_range * padding_factor
+        # y_max += y_range * padding_factor
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:322]: is this worth implementing?  If not, maybe we can remove?

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.vue [hunk 11]
```diff
@@ -2,6 +2,9 @@
   <j-tray-plugin
     :description="docs_description || 'Queries an area encompassed by the viewer using a specified catalog and marks all the objects found within the area.'"
     :link="docs_link || 'https://jdaviz.readthedocs.io/en/'+vdocs+'/'+config+'/plugins.html#catalog-search'"
+    :uses_active_status="uses_active_status"
+    @plugin-ping="plugin_ping($event)"
+    :keept_active_sync="keep_active"
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.vue:7]:
```suggestion
    :keep_active_sync="keep_active"
```

**Review by kecnry — APPROVED** [review]

**Comment by pllim** [comment]
> I'm struggling a bit with importing a custom table.

My conclusion on this problem from my own testing of the plugin is that roundtrip is currently not possible. When you export, the RA and Dec are separate columns. So to re-import it, I have to manually combine the two columns into SkyCoord first, add that as a new `sky_centroid` column, and save that updated table out again myself. Only then, I can import the table back in. I think there is a JIRA ticket about this somewhere but I don't know how to dig it back out.
